### DISCUSSION TITLE
errors when pulling or cloning lots of repos

### DISCUSSION THREAD

**Discussion by breed** [community/community#111422] [GitHub Education]
### Hi everyone,

i have problems when cloning or pulling more than 10 repos at a time like:

    gh classroom pull student-repos

after a few repos, get the following error:

    failed to run gh: kex_exchange_identification: read: Connection reset by peer

it appears to be due to some rate limiting. if i use a script to pull each repo instead, i can put a 10 second sleep between each pull and everything will work.

i'd like to use gh since the commands are much nicer than doing raw git, but with a class of 40 students, i keep hitting this problem.

has anyone else seen/resolved this?

**Comment by ngljcb** [comment]
Encountering rate limiting problems while trying to clone or pull multiple repositories from GitHub is a common issue, especially when dealing with a large number of repositories within a short timeframe. The error message you're seeing (kex_exchange_identification: read: Connection reset by peer) indicates a possible network issue or a connection reset by the server, which often occurs due to rate limiting measures being triggered.

To overcome these rate limiting challenges, one solution is to use a script that introduces a delay between each pull operation. This helps space out the requests, preventing them from overwhelming the server and triggering rate limiting mechanisms.

Alternatively, you could utilize GitHub's GraphQL API to fetch repository information instead of relying solely on tools like gh or raw Git commands. This approach grants you more control over the requests and enables you to implement strategies for handling rate limiting errors more effectively. By batching repository requests and implementing retry logic, you can manage rate limiting issues more gracefully and ensure successful repository pulls for your entire class.

In summary, employing strategies such as introducing delays between pull operations or utilizing GitHub's GraphQL API can help mitigate rate limiting problems and facilitate the smooth retrieval of repositories for your class without encountering these issues.
